### PR TITLE
feat: Add support for using watchman(1) instead of walking filesystem

### PR DESCRIPTION
**Pull request by KAction**
Anecodically, for project having 4k files and 7 different formatters, this change can reduce time it takes for treefmt(1) to figure out that it has nothing to do from ~350ms down to ~55ms, out of which getting response from watchman daemon takes ~35ms, so there is room for further improvements.

If "WATCHMAN_SOCK" environment variable is not set and "watchman(1)" command are not available, "treefmt" transparently falls back on walking the file system. Setting the environment variable is highly recommended, since is avoids subprocess invocation.

Ref: https://github.com/numtide/treefmt/issues/261

Regarding further optimizations, this is what I think can be improved::

```
[DEBUG] load cache: 8.36ms (Δ 6.85ms)
...
[DEBUG] watchman query: 43.73ms (Δ 35.02ms)
[DEBUG] post-process watchman list: 50.55ms (Δ 6.82ms)
...
```

### CODE REVIEW

File: src/engine.rs [hunk 1]
```diff
@@ -21,9 +26,43 @@ pub enum DisplayType {
...
+
+// Here "path" is the absolute path that matches "includes" patterns and was changed, but that file
+// might be need to be excluded since its parent directory matches "excludes" patterns. Hence, we
+// need to iterate parent directories.
+//
+// Filesystem walker does not need this, since it would not descend into excluded directory in the
+// first place to discover a changed file that matches "includes" pattern.
+//
+// FIXME: There should be a cleaner solution.
+fn path_excluded(gs: &GlobSet, path: &PathBuf) -> bool {
```
Comment (KAction) [inline src/engine.rs:45]: This one is really inefficient. For nixpkgs, I get following timings:
```
[DEBUG] watchman query: 404.46ms (Δ 120.62ms)
[DEBUG] post-process watchman list: 601.97ms (Δ 197.52ms)
```

**Comment by KAction** [comment]
As for excludes, I think I can at least optimize happy path and change `excludes: Option<GlobSet>`.

**Comment by KAction** [comment]
Updated the pull request. Previous version misused watchman, causing it to result way too much results.
With this version, whole nixpkgs is down to 12ms for no-op, my $dayjob repository is down to 8.5ms.

**Comment by KAction** [comment]
```
error (ignored): error: SQLite database '/home/garnix/.cache/nix/eval-cache-v5/6de50559647aa37e20a3fd649274afe85c9def927c4d55696d8a84953e4ccaa3.sqlite' is busy
```

Should I be concerned?

File: src/command/format.rs [hunk 2]
```diff
@@ -49,6 +52,25 @@ pub fn format_cmd(
         paths
     );
 
+    let client = match Connector::new().connect().await {
+        Err(e) => {
+            warn!(
+                "watchman is not available (err = {:?}), falling back on stat(2)",
+                e
+            );
```
Comment (zimbatm) [inline src/command/format.rs:60]: Most users should be able to use treefmt without `watchman`, so having this as a warning isn't the best user experience for them.

I propose three modes: "auto" (default) tries watchman and silently falls back on "stat," "stat" is stat, and "watchman" errors if watchman is not available.

Then, introduce a CLI flag and/or project-level global config to allow switching between modes.

File: src/command/format.rs [hunk 3]
```diff
@@ -49,6 +52,25 @@ pub fn format_cmd(
...
+            None
+        }
+        Ok(c) => {
+            // This is important. Subprocess wastes ~20ms.
+            if !std::env::var("WATCHMAN_SOCK").is_ok() {
```
Comment (zimbatm) [inline src/command/format.rs:65]: this needs to be documented, to explain how to set it

File: src/formatter.rs [hunk 4]
```diff
@@ -191,7 +201,10 @@ impl Formatter {
             options: cfg.options.clone(),
             work_dir,
             includes,
+            // TODO: Avoid clone().
```
Comment (zimbatm) [inline src/formatter.rs:204]: the list is generally super small, I don't think the TODO needs to stay.

**Review by zimbatm — COMMENTED** [review]
hey, I finally had the time to take a closer look and actually run it, and it's great. The only thing missing is to fix the default user experience if watchman is not installed, and a bit of documentation on how to use that feature.

**Comment by KAction** [comment]
I think I addressed all suggestions it latest revision.

**Pull request closed without merging**